From 2806eeb93b3e9f84f8d3255c523a1868d1406846 Mon Sep 17 00:00:00 2001
From: Bart Van Assche <bvanassche@acm.org>
Date: Thu, 26 Dec 2019 00:28:14 +0000
Subject: [PATCH] qla2x00t-32gbit: Correctly retrieve and interpret active
 flash region

ISP27XX/28XX supports multiple flash regions. This patch fixes issue where
active flash region was not interpreted correctly during secure flash
update process.

Fixes: 5fa8774c7f38c ("scsi: qla2xxx: Add 28xx flash primary/secondary status/image mechanism")
Cc: stable@vger.kernel.org
Link: https://lore.kernel.org/r/20191203223657.22109-2-hmadhani@marvell.com
Signed-off-by: Michael Hernandez <mhernandez@marvell.com>
Signed-off-by: Himanshu Madhani <hmadhani@marvell.com>
Signed-off-by: Martin K. Petersen <martin.petersen@oracle.com>
[mkp: typo]

[ commit 4e71dcae0c4cd1e9d19b8b3d80214a4bcdca5a42 upstream ]


git-svn-id: http://svn.code.sf.net/p/scst/svn/trunk@8727 d57e44dd-8a1f-0410-8b47-8ef2f437770f
---
 qla2x00t-32gbit/qla_attr.c | 1 +
 qla2x00t-32gbit/qla_bsg.c  | 2 +-
 qla2x00t-32gbit/qla_sup.c  | 6 +++---
 3 files changed, 5 insertions(+), 4 deletions(-)

Index: scst-3.4.x/qla2x00t-32gbit/qla_attr.c
===================================================================
--- scst-3.4.x.orig/qla2x00t-32gbit/qla_attr.c
+++ scst-3.4.x/qla2x00t-32gbit/qla_attr.c
@@ -176,6 +176,7 @@ qla2x00_sysfs_read_nvram(struct file *fi
 
 	faddr = ha->flt_region_nvram;
 	if (IS_QLA28XX(ha)) {
+		qla28xx_get_aux_images(vha, &active_regions);
 		if (active_regions.aux.vpd_nvram == QLA27XX_SECONDARY_IMAGE)
 			faddr = ha->flt_region_nvram_sec;
 	}
Index: scst-3.4.x/qla2x00t-32gbit/qla_bsg.c
===================================================================
--- scst-3.4.x.orig/qla2x00t-32gbit/qla_bsg.c
+++ scst-3.4.x/qla2x00t-32gbit/qla_bsg.c
@@ -2559,7 +2559,7 @@ qla2x00_get_flash_image_status(struct bs
 	struct qla_active_regions regions = { };
 	struct active_regions active_regions = { };
 
-	qla28xx_get_aux_images(vha, &active_regions);
+	qla27xx_get_active_image(vha, &active_regions);
 	regions.global_image = active_regions.global;
 
 	if (IS_QLA28XX(ha)) {
Index: scst-3.4.x/qla2x00t-32gbit/qla_sup.c
===================================================================
--- scst-3.4.x.orig/qla2x00t-32gbit/qla_sup.c
+++ scst-3.4.x/qla2x00t-32gbit/qla_sup.c
@@ -847,15 +847,15 @@ qla2xxx_get_flt_info(scsi_qla_host_t *vh
 				ha->flt_region_img_status_pri = start;
 			break;
 		case FLT_REG_IMG_SEC_27XX:
-			if (IS_QLA27XX(ha) && !IS_QLA28XX(ha))
+			if (IS_QLA27XX(ha) || IS_QLA28XX(ha))
 				ha->flt_region_img_status_sec = start;
 			break;
 		case FLT_REG_FW_SEC_27XX:
-			if (IS_QLA27XX(ha) && !IS_QLA28XX(ha))
+			if (IS_QLA27XX(ha) || IS_QLA28XX(ha))
 				ha->flt_region_fw_sec = start;
 			break;
 		case FLT_REG_BOOTLOAD_SEC_27XX:
-			if (IS_QLA27XX(ha) && !IS_QLA28XX(ha))
+			if (IS_QLA27XX(ha) || IS_QLA28XX(ha))
 				ha->flt_region_boot_sec = start;
 			break;
 		case FLT_REG_AUX_IMG_PRI_28XX:
